teps/0033-tekton-feature-gates.md

```diff
@@ -145,6 +160,393 @@ alpha features. This is not a requirement for Tekton, because even in case of
 a managed Tekton, users can be prevented access to the `tekton-pipelines`
 namespaces, where the feature flag config map is hosted.
 
+## Proposal
+
+We will tackle separately features that are controlled via new fields in the API
+and features that are only behavioral. (The alternative
+[one feature flag per new alpha field](#one-feature-flag-per-new-alpha-field) describes
+an alternative where we use the same approach for both.)
+
+In both cases we want to control access with
+[a feature flag](https://github.com/tektoncd/pipeline/blob/master/docs/install.md#customizing-the-pipelines-controller-behavior).
+
+* For new API fields, will will gate access to all of them behind two feature flags
+  called `enable-alpha-api-fields` and `enable-beta-api-fields`.
+  ([Design details: new api fields](#new-api-fields).)
+* For changes in behavior which are not configurable via API fields in our CRDs,
+  each feature will be gated by its own unique feature flag.
+  ([Design details: changes in behavior](#changes-in-behavior).)
+
+
+### Risks and Mitigations
+
+See the pros and cons sections in the [design details](#design-details)
+below.
+
+## Design Details
+
+### New API fields
+
+We gate access to all new API fields with two new
+[feature flags](https://github.com/tektoncd/pipeline/blob/master/docs/install.md#customizing-the-pipelines-controller-behavior) called `enable-alpha-api-fields` and `enable-beta-api-fields`
+which will default to false.
+
+For example:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: feature-flags
+data:
+  enable-alpha-api-fields: "true"
+  enable-beta-api-fields: "true"
+```
+
+This allows administrators to opt into allowing their users to use alpha and beta fields.
+
+_Note that we will only need the `enable-beta-api-fields` flag once we have v1 types in Tekton,
+at the moment the highest version we have is beta, so we will initially only need
+`enable-alpha-api-fields`._
```

Comment:
Here, `alpha` and `beta` are referring to individual features and not overall type version which is represented using `APIVersion`.

The highest `APIVersion` we have is `v1beta1`. When I am introducing a new feature in `v1beta1`, it will considered as `alpha` level support. After many iterations and improvements, we can upgrade that feature to `beta` in the same API `v1beta1` but not to `GA`.

A year from now, our highest `APIVersion` would be `v1` (may be). I am guessing this proposal allows us to introduce an `alpha` feature in `v1`. In that case, irrespective of what `APIVersion` we are on, we will need to serve all lower possible levels, i.e. for `v1` types, we will need `alpha` and `beta` flags.

Reply:
Yes. in a year, our `APIVersion` would be `v1`, hopefully `v1alpha1` would be gone already, and `v1beta1` on it's way to be gone too. So yes, this proposal is about new alpha/beta fields in `v1` types.

Reply:
> So yes, this proposal is about new alpha/beta fields in v1 types.

it is possible that we could introduce new types in the future which would start out at v1alpha1 as well, so it'd probably be best if our solution could work for those types as well